File: THExtensions/THNSMenuExtensions.m
// THNSMenuExtensions.m

#import "THNSMenuExtensions.h"

//--------------------------------------------------------------------------------------------------------------------------------------------
@implementation NSMenuItem (THNSMenuExtensions)

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title target:(id)target action:(SEL)action representedObject:(id)representedObject tag:(NSInteger)tag isEnabled:(BOOL)isEnabled
{
	NSMenuItem *result=[[self alloc] initWithTitle:title!=nil?title:@"" action:action keyEquivalent:@""];
	result.target=target;
	result.representedObject=representedObject;
	result.tag=tag;
	[result setEnabled:isEnabled];
	return result;
}

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title target:(id)target action:(SEL)action representedObject:(id)representedObject isEnabled:(BOOL)isEnabled
{
	NSMenuItem *result=[[self alloc] initWithTitle:title!=nil?title:@"" action:action keyEquivalent:@""];
	result.target=target;
	result.representedObject=representedObject;
	[result setEnabled:isEnabled];
	return result;
}

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title target:(id)target action:(SEL)action tag:(NSInteger)tag
{
	NSMenuItem *result=[[self alloc] initWithTitle:title!=nil?title:@"" action:action keyEquivalent:@""];
	result.target=target;
	result.tag=tag;
	return result;
}

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title target:(id)target action:(SEL)action tag:(NSInteger)tag isEnabled:(BOOL)isEnabled
{
	NSMenuItem *result=[self th_menuItemWithTitle:title target:target action:action tag:tag];
	[result setEnabled:isEnabled];
	return result;
}

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title target:(id)target action:(SEL)action
{
	NSMenuItem *result=[[self alloc] initWithTitle:title!=nil?title:@"" action:action keyEquivalent:@""];
	result.target=target;
	result.action=action;
	return result;
}

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title tag:(NSInteger)tag
{
	NSMenuItem *result=[[self alloc] initWithTitle:title!=nil?title:@"" action:nil keyEquivalent:@""];
	result.tag=tag;
	return result;
}

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title tag:(NSInteger)tag isEnabled:(BOOL)isEnabled
{
	NSMenuItem *result=[[self alloc] initWithTitle:title!=nil?title:@"" action:nil keyEquivalent:@""];
	result.tag=tag;
	[result setEnabled:isEnabled];
	return result;
}

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title tag:(NSInteger)tag representedObject:(id)representedObject
{
	NSMenuItem *result=[[self alloc] initWithTitle:title!=nil?title:@"" action:nil keyEquivalent:@""];
	result.tag=tag;
	result.representedObject=representedObject;
	return result;
}

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title representedObject:(id)representedObject
{
	NSMenuItem *result=[[self alloc] initWithTitle:title!=nil?title:@"" action:nil keyEquivalent:@""];
	result.representedObject=representedObject;
	return result;
}

+ (NSMenuItem*)th_menuItemWithTitle:(NSString*)title representedObject:(id)representedObject isEnabled:(BOOL)isEnabled
{
	NSMenuItem *result=[[self alloc] initWithTitle:title!=nil?title:@"" action:nil keyEquivalent:@""];
	result.representedObject=representedObject;
	[result setEnabled:isEnabled];
	return result;
}

@end
//--------------------------------------------------------------------------------------------------------------------------------------------

File: THExtensions/THNSMenuExtensions.swift
// THNSMenuExtensions.swift

import Cocoa

//--------------------------------------------------------------------------------------------------------------------------------------------
extension NSMenu {

	@objc convenience init(title: String?, delegate: NSMenuDelegate, autoenablesItems: Bool) {
		if let title = title {
			self.init(title: title)
		}
		else {
			self.init()
		}
		
		self.delegate = delegate
		self.autoenablesItems = autoenablesItems
	}

	@objc func th_lastItem() -> NSMenuItem? {
		self.numberOfItems > 0 ? self.item(at: self.numberOfItems - 1) : nil
	}

}
//--------------------------------------------------------------------------------------------------------------------------------------------


//--------------------------------------------------------------------------------------------------------------------------------------------
extension NSMenuItem {

	@objc convenience init(title: String, target: AnyObject? = nil, action: Selector? = nil, representedObject: Any? = nil, tag: Int = 0, enabled: Bool = true, submenu: NSMenu? = nil) {
		self.init(title: title, action: action, keyEquivalent: "")

		if let target = target {
			self.target = target
		}
		
		self.isEnabled = enabled
		
		if let representedObject = representedObject {
			self.representedObject = representedObject
		}
		
		if tag != 0 {
			self.tag = tag
		}

		if let submenu = submenu {
			self.submenu = submenu
		}
	}

}
//--------------------------------------------------------------------------------------------------------------------------------------------


//--------------------------------------------------------------------------------------------------------------------------------------------
class THMenuItem: NSMenuItem {
	var actionBlock: (() -> Void)!

	convenience init(title: String, representedObject: Any? = nil, tag: Int = 0, enabled: Bool = true, block: @escaping () -> Void) {
		self.init(title: title, target: nil, action: #selector(mi_action), representedObject: representedObject, tag: tag, enabled: enabled)
		self.target = self
		self.actionBlock = block
	}

	@objc func mi_action(_ sender: NSMenuItem) {
		actionBlock()
	}

}
//--------------------------------------------------------------------------------------------------------------------------------------------
